File: ios/SkyCast/Data/Mapper/WeatherMapper.swift
import Foundation

/// Every conversion between wire/storage shapes and domain models.
///
/// These are **pure functions with no dependencies**, which makes them the cheapest and
/// highest-value tests in the project, see `WeatherMapperTests`. Keeping all mapping here
/// means an OpenWeather response change touches this file and nothing else.
enum WeatherMapper {
    // MARK: - Remote → Domain

    static func weather(
        from dto: CurrentWeatherDTO,
        locationID: Int64,
        locationName: String,
        cachedAt: Date
    )
        -> Weather
    {
        // OpenWeather always sends at least one entry, but tolerating its absence keeps a
        // malformed response from taking the app down.
        let primary = dto.weather.first

        return Weather(
            locationID: locationID,
            locationName: locationName.isEmpty ? dto.cityName : locationName,
            condition: WeatherCondition.fromOpenWeatherID(primary?.id ?? 0),
            description: (primary?.description ?? "").capitalizedFirstLetter,
            iconCode: primary?.icon ?? "",
            temperatureCelsius: dto.main.temperature,
            feelsLikeCelsius: dto.main.feelsLike,
            minTemperatureCelsius: dto.main.temperatureMin,
            maxTemperatureCelsius: dto.main.temperatureMax,
            humidityPercent: dto.main.humidity,
            pressureHpa: dto.main.pressure,
            windSpeedMetresPerSecond: dto.wind.speed,
            windDirectionDegrees: dto.wind.degrees,
            cloudinessPercent: dto.clouds.cloudinessPercent,
            visibilityMetres: dto.visibility,
            sunrise: Date(timeIntervalSince1970: TimeInterval(dto.system.sunriseEpochSeconds)),
            sunset: Date(timeIntervalSince1970: TimeInterval(dto.system.sunsetEpochSeconds)),
            observedAt: Date(timeIntervalSince1970: TimeInterval(dto.observedAtEpochSeconds)),
            cachedAt: cachedAt
        )
    }

    /// Groups the flat 3-hourly list into calendar days.
    ///
    /// Days are keyed in the **location's** timezone, not the device's: a forecast for
    /// Male' viewed from London must still be grouped by Maldivian days.
    static func forecast(
        from dto: ForecastResponseDTO,
        locationID: Int64,
        locationName: String,
        cachedAt: Date
    )
        -> Forecast
    {
        var calendar = Calendar(identifier: .gregorian)
        // `.gmt` rather than `TimeZone(secondsFromGMT: 0)!`, same value, no force unwrap.
        // A malformed offset from the API must degrade to GMT, not trap.
        calendar.timeZone = TimeZone(secondsFromGMT: dto.city.timezoneOffsetSeconds) ?? .gmt

        let readings = dto.readings.map { reading -> DatedReading in
            let time = Date(timeIntervalSince1970: TimeInterval(reading.timeEpochSeconds))
            let hourly = HourlyForecast(
                time: time,
                condition: WeatherCondition.fromOpenWeatherID(reading.weather.first?.id ?? 0),
                iconCode: reading.weather.first?.icon ?? "",
                temperatureCelsius: reading.main.temperature,
                precipitationProbability: reading.precipitationProbability,
                windSpeedMetresPerSecond: reading.wind.speed
            )
            return DatedReading(day: calendar.startOfDay(for: time), hourly: hourly, dto: reading)
        }

        let grouped = Dictionary(grouping: readings, by: \.day)

        let days: [ForecastDay] = grouped.keys.sorted().compactMap { day in
            guard let entries = grouped[day], !entries.isEmpty else { return nil }
            let representative = representativeEntry(from: entries, calendar: calendar)

            return ForecastDay(
                date: day,
                condition: representative.hourly.condition,
                description: (representative.dto.weather.first?.description ?? "").capitalizedFirstLetter,
                iconCode: representative.hourly.iconCode,
                minTemperatureCelsius: entries.map(\.dto.main.temperatureMin).min() ?? 0,
                maxTemperatureCelsius: entries.map(\.dto.main.temperatureMax).max() ?? 0,
                precipitationProbability: entries.map(\.hourly.precipitationProbability).max() ?? 0,
                hourly: entries.map(\.hourly).sorted { $0.time < $1.time }
            )
        }

        return Forecast(
            locationID: locationID,
            locationName: locationName.isEmpty ? dto.city.name : locationName,
            days: days,
            cachedAt: cachedAt
        )
    }

    static func searchResult(from dto: GeocodingResultDTO) -> LocationSearchResult {
        LocationSearchResult(
            name: dto.name,
            countryCode: dto.country,
            state: dto.state,
            latitude: dto.latitude,
            longitude: dto.longitude
        )
    }

    // MARK: - Domain → Local

    static func persistentWeather(from weather: Weather) -> PersistentWeather {
        PersistentWeather(
            locationID: weather.locationID,
            locationName: weather.locationName,
            conditionRawValue: weather.condition.rawValue,
            weatherDescription: weather.description,
            iconCode: weather.iconCode,
            temperatureCelsius: weather.temperatureCelsius,
            feelsLikeCelsius: weather.feelsLikeCelsius,
            minTemperatureCelsius: weather.minTemperatureCelsius,
            maxTemperatureCelsius: weather.maxTemperatureCelsius,
            humidityPercent: weather.humidityPercent,
            pressureHpa: weather.pressureHpa,
            windSpeedMetresPerSecond: weather.windSpeedMetresPerSecond,
            windDirectionDegrees: weather.windDirectionDegrees,
            cloudinessPercent: weather.cloudinessPercent,
            visibilityMetres: weather.visibilityMetres,
            sunrise: weather.sunrise,
            sunset: weather.sunset,
            observedAt: weather.observedAt,
            cachedAt: weather.cachedAt
        )
    }

    static func persistentReadings(from forecast: Forecast) -> [PersistentForecastReading] {
        forecast.days.flatMap { day in
            day.hourly.map { hour in
                PersistentForecastReading(
                    locationID: forecast.locationID,
                    locationName: forecast.locationName,
                    time: hour.time,
                    conditionRawValue: hour.condition.rawValue,
                    weatherDescription: day.description,
                    iconCode: hour.iconCode,
                    temperatureCelsius: hour.temperatureCelsius,
                    precipitationProbability: hour.precipitationProbability,
                    windSpeedMetresPerSecond: hour.windSpeedMetresPerSecond,
                    cachedAt: forecast.cachedAt
                )
            }
        }
    }

    static func persistentLocation(from location: SavedLocation) -> PersistentSavedLocation {
        PersistentSavedLocation(
            id: location.id,
            name: location.name,
            countryCode: location.countryCode,
            state: location.state,
            latitude: location.latitude,
            longitude: location.longitude,
            sortOrder: location.sortOrder,
            isPrimary: location.isPrimary
        )
    }

    // MARK: - Local → Domain

    static func weather(from model: PersistentWeather) -> Weather {
        Weather(
            locationID: model.locationID,
            locationName: model.locationName,
            condition: WeatherCondition(rawValue: model.conditionRawValue) ?? .unknown,
            description: model.weatherDescription,
            iconCode: model.iconCode,
            temperatureCelsius: model.temperatureCelsius,
            feelsLikeCelsius: model.feelsLikeCelsius,
            minTemperatureCelsius: model.minTemperatureCelsius,
            maxTemperatureCelsius: model.maxTemperatureCelsius,
            humidityPercent: model.humidityPercent,
            pressureHpa: model.pressureHpa,
            windSpeedMetresPerSecond: model.windSpeedMetresPerSecond,
            windDirectionDegrees: model.windDirectionDegrees,
            cloudinessPercent: model.cloudinessPercent,
            visibilityMetres: model.visibilityMetres,
            sunrise: model.sunrise,
            sunset: model.sunset,
            observedAt: model.observedAt,
            cachedAt: model.cachedAt
        )
    }

    /// Rebuilds the day grouping from cached records, in the device's timezone.
    static func forecast(from readings: [PersistentForecastReading]) -> Forecast? {
        guard let first = readings.first else { return nil }

        let calendar = Calendar.current
        let grouped = Dictionary(grouping: readings) { calendar.startOfDay(for: $0.time) }

        let days: [ForecastDay] = grouped.keys.sorted().compactMap { day in
            guard let entries = grouped[day], !entries.isEmpty else { return nil }

            let representative = entries.min { lhs, rhs in
                let lhsDistance = abs(calendar.component(.hour, from: lhs.time) - Constants.middayHour)
                let rhsDistance = abs(calendar.component(.hour, from: rhs.time) - Constants.middayHour)
                return lhsDistance < rhsDistance
            } ?? entries[0]

            return ForecastDay(
                date: day,
                condition: WeatherCondition(rawValue: representative.conditionRawValue) ?? .unknown,
                description: representative.weatherDescription,
                iconCode: representative.iconCode,
                minTemperatureCelsius: entries.map(\.temperatureCelsius).min() ?? 0,
                maxTemperatureCelsius: entries.map(\.temperatureCelsius).max() ?? 0,
                precipitationProbability: entries.map(\.precipitationProbability).max() ?? 0,
                hourly: entries
                    .sorted { $0.time < $1.time }
                    .map { reading in
                        HourlyForecast(
                            time: reading.time,
                            condition: WeatherCondition(rawValue: reading.conditionRawValue) ?? .unknown,
                            iconCode: reading.iconCode,
                            temperatureCelsius: reading.temperatureCelsius,
                            precipitationProbability: reading.precipitationProbability,
                            windSpeedMetresPerSecond: reading.windSpeedMetresPerSecond
                        )
                    }
            )
        }

        return Forecast(
            locationID: first.locationID,
            locationName: first.locationName,
            days: days,
            cachedAt: first.cachedAt
        )
    }

    static func location(from model: PersistentSavedLocation) -> SavedLocation {
        SavedLocation(
            id: model.id,
            name: model.name,
            countryCode: model.countryCode,
            state: model.state,
            latitude: model.latitude,
            longitude: model.longitude,
            sortOrder: model.sortOrder,
            isPrimary: model.isPrimary
        )
    }

    // MARK: - Internals

    /// The reading nearest local noon best characterises a whole day; the 03:00 reading
    /// would make every day look clear and cold.
    private static func representativeEntry(
        from entries: [DatedReading],
        calendar: Calendar
    )
        -> DatedReading
    {
        entries.min { lhs, rhs in
            let lhsDistance = abs(calendar.component(.hour, from: lhs.hourly.time) - Constants.middayHour)
            let rhsDistance = abs(calendar.component(.hour, from: rhs.hourly.time) - Constants.middayHour)
            return lhsDistance < rhsDistance
        } ?? entries[0]
    }

    /// A forecast reading paired with the calendar day it falls in, in the *location's*
    /// timezone.
    ///
    /// A named type rather than a tuple: a three-member tuple is at the limit of what stays
    /// readable, and `entries.map(\.dto.main.temperatureMin)` is far clearer than `.2.main…`.
    private struct DatedReading {
        let day: Date
        let hourly: HourlyForecast
        let dto: ForecastReadingDTO
    }

    private enum Constants {
        static let middayHour = 12
    }
}

private extension String {
    /// OpenWeather returns lowercase descriptions ("light rain"); we display them
    /// sentence-cased without touching the rest of the string.
    var capitalizedFirstLetter: String {
        guard let first else { return self }
        return first.uppercased() + dropFirst()
    }
}

File: android/core/data/src/main/kotlin/com/nauhaan/skycast/data/mapper/WeatherMapper.kt
package com.nauhaan.skycast.data.mapper

import com.nauhaan.skycast.data.local.entity.CachedForecastReadingEntity
import com.nauhaan.skycast.data.local.entity.CachedWeatherEntity
import com.nauhaan.skycast.data.local.entity.SavedLocationEntity
import com.nauhaan.skycast.data.remote.dto.CurrentWeatherDto
import com.nauhaan.skycast.data.remote.dto.ForecastResponseDto
import com.nauhaan.skycast.data.remote.dto.GeocodingResultDto
import com.nauhaan.skycast.domain.model.Forecast
import com.nauhaan.skycast.domain.model.ForecastDay
import com.nauhaan.skycast.domain.model.HourlyForecast
import com.nauhaan.skycast.domain.model.LocationSearchResult
import com.nauhaan.skycast.domain.model.SavedLocation
import com.nauhaan.skycast.domain.model.Weather
import com.nauhaan.skycast.domain.model.WeatherCondition
import java.time.Instant
import java.time.ZoneId
import java.time.ZoneOffset

/**
 * Every conversion between wire/storage shapes and domain models.
 *
 * These are **pure functions with no dependencies**, which makes them the cheapest
 * and highest-value unit tests in the project; see `WeatherMapperTest`.
 *
 * Keeping all mapping here means an OpenWeather response change touches this file
 * and nothing else.
 */
object WeatherMapper {
    // ── Remote → Domain ────────────────────────────────────────────────────

    fun CurrentWeatherDto.toDomain(locationId: Long, locationName: String, cachedAt: Instant): Weather {
        // OpenWeather always sends at least one entry, but defaulting keeps a
        // malformed response from crashing the app.
        val primary = weather.firstOrNull()
        return Weather(
            locationId = locationId,
            locationName = locationName.ifBlank { cityName },
            condition = WeatherCondition.fromOpenWeatherId(primary?.id ?: 0),
            description = primary?.description?.replaceFirstChar(Char::uppercase).orEmpty(),
            iconCode = primary?.icon.orEmpty(),
            temperatureCelsius = main.temperature,
            feelsLikeCelsius = main.feelsLike,
            minTemperatureCelsius = main.temperatureMin,
            maxTemperatureCelsius = main.temperatureMax,
            humidityPercent = main.humidity,
            pressureHpa = main.pressure,
            windSpeedMetresPerSecond = wind.speed,
            windDirectionDegrees = wind.degrees,
            cloudinessPercent = clouds.cloudinessPercent,
            visibilityMetres = visibility,
            sunrise = Instant.ofEpochSecond(system.sunriseEpochSeconds),
            sunset = Instant.ofEpochSecond(system.sunsetEpochSeconds),
            observedAt = Instant.ofEpochSecond(observedAtEpochSeconds),
            cachedAt = cachedAt,
            zoneOffset = safeOffset(timezoneOffsetSeconds),
        )
    }

    /**
     * Groups the flat 3-hourly list into calendar days.
     *
     * Days are keyed in the **location's** timezone, not the device's: a forecast for
     * Male' viewed from London must still be grouped by Maldivian days.
     */
    fun ForecastResponseDto.toDomain(locationId: Long, locationName: String, cachedAt: Instant): Forecast {
        val offset = safeOffset(city.timezoneOffsetSeconds)
        val zone: ZoneId = offset

        val days =
            readings
                .groupBy { Instant.ofEpochSecond(it.timeEpochSeconds).atZone(zone).toLocalDate() }
                .toSortedMap()
                .map { (date, dayReadings) ->
                    // The reading nearest midday best represents the day as a whole; the
                    // 03:00 reading would make every day look clear and cold.
                    val representative =
                        dayReadings.minByOrNull { reading ->
                            val hour = Instant.ofEpochSecond(reading.timeEpochSeconds).atZone(zone).hour
                            kotlin.math.abs(hour - MIDDAY_HOUR)
                        } ?: dayReadings.first()
                    val primary = representative.weather.firstOrNull()

                    ForecastDay(
                        date = date,
                        condition = WeatherCondition.fromOpenWeatherId(primary?.id ?: 0),
                        description = primary?.description?.replaceFirstChar(Char::uppercase).orEmpty(),
                        iconCode = primary?.icon.orEmpty(),
                        minTemperatureCelsius = dayReadings.minOf { it.main.temperatureMin },
                        maxTemperatureCelsius = dayReadings.maxOf { it.main.temperatureMax },
                        precipitationProbability = dayReadings.maxOf { it.precipitationProbability },
                        hourly = dayReadings.map { reading ->
                            HourlyForecast(
                                time = Instant.ofEpochSecond(reading.timeEpochSeconds),
                                condition = WeatherCondition.fromOpenWeatherId(
                                    reading.weather.firstOrNull()?.id ?: 0,
                                ),
                                iconCode = reading.weather
                                    .firstOrNull()
                                    ?.icon
                                    .orEmpty(),
                                temperatureCelsius = reading.main.temperature,
                                precipitationProbability = reading.precipitationProbability,
                                windSpeedMetresPerSecond = reading.wind.speed,
                            )
                        },
                    )
                }

        return Forecast(
            locationId = locationId,
            locationName = locationName.ifBlank { city.name },
            days = days,
            cachedAt = cachedAt,
            zoneOffset = offset,
        )
    }

    fun GeocodingResultDto.toDomain(): LocationSearchResult = LocationSearchResult(
        name = name,
        countryCode = country,
        state = state,
        latitude = latitude,
        longitude = longitude,
    )

    // ── Domain → Local ─────────────────────────────────────────────────────

    fun Weather.toEntity(): CachedWeatherEntity = CachedWeatherEntity(
        locationId = locationId,
        locationName = locationName,
        conditionId = condition.ordinal,
        description = description,
        iconCode = iconCode,
        temperatureCelsius = temperatureCelsius,
        feelsLikeCelsius = feelsLikeCelsius,
        minTemperatureCelsius = minTemperatureCelsius,
        maxTemperatureCelsius = maxTemperatureCelsius,
        humidityPercent = humidityPercent,
        pressureHpa = pressureHpa,
        windSpeedMetresPerSecond = windSpeedMetresPerSecond,
        windDirectionDegrees = windDirectionDegrees,
        cloudinessPercent = cloudinessPercent,
        visibilityMetres = visibilityMetres,
        sunriseEpochSeconds = sunrise.epochSecond,
        sunsetEpochSeconds = sunset.epochSecond,
        observedAtEpochSeconds = observedAt.epochSecond,
        cachedAtEpochSeconds = cachedAt.epochSecond,
        timezoneOffsetSeconds = zoneOffset.totalSeconds,
    )

    fun Forecast.toEntities(): List<CachedForecastReadingEntity> = days.flatMap { day ->
        day.hourly.map { hour ->
            CachedForecastReadingEntity(
                locationId = locationId,
                locationName = locationName,
                timeEpochSeconds = hour.time.epochSecond,
                conditionId = hour.condition.ordinal,
                description = day.description,
                iconCode = hour.iconCode,
                temperatureCelsius = hour.temperatureCelsius,
                precipitationProbability = hour.precipitationProbability,
                windSpeedMetresPerSecond = hour.windSpeedMetresPerSecond,
                cachedAtEpochSeconds = cachedAt.epochSecond,
                timezoneOffsetSeconds = zoneOffset.totalSeconds,
            )
        }
    }

    // ── Local → Domain ─────────────────────────────────────────────────────

    fun CachedWeatherEntity.toDomain(): Weather = Weather(
        locationId = locationId,
        locationName = locationName,
        condition = WeatherCondition.entries.getOrElse(conditionId) { WeatherCondition.UNKNOWN },
        description = description,
        iconCode = iconCode,
        temperatureCelsius = temperatureCelsius,
        feelsLikeCelsius = feelsLikeCelsius,
        minTemperatureCelsius = minTemperatureCelsius,
        maxTemperatureCelsius = maxTemperatureCelsius,
        humidityPercent = humidityPercent,
        pressureHpa = pressureHpa,
        windSpeedMetresPerSecond = windSpeedMetresPerSecond,
        windDirectionDegrees = windDirectionDegrees,
        cloudinessPercent = cloudinessPercent,
        visibilityMetres = visibilityMetres,
        sunrise = Instant.ofEpochSecond(sunriseEpochSeconds),
        sunset = Instant.ofEpochSecond(sunsetEpochSeconds),
        observedAt = Instant.ofEpochSecond(observedAtEpochSeconds),
        cachedAt = Instant.ofEpochSecond(cachedAtEpochSeconds),
        zoneOffset = safeOffset(timezoneOffsetSeconds),
    )

    /**
     * Rebuilds the day grouping from cached rows.
     *
     * Groups in the **location's** timezone, exactly as the remote path above does. Using the
     * device's zone here instead would make a day's boundary depend on
     * which path served the read, so the `epochDay` captured from a freshly-fetched list could
     * fail to resolve against the same forecast reloaded from cache.
     */
    fun List<CachedForecastReadingEntity>.toDomainForecast(): Forecast? {
        val first = firstOrNull() ?: return null
        val offset = safeOffset(first.timezoneOffsetSeconds)
        val zone: ZoneId = offset

        val days =
            groupBy { Instant.ofEpochSecond(it.timeEpochSeconds).atZone(zone).toLocalDate() }
                .toSortedMap()
                .map { (date, readings) ->
                    val representative =
                        readings.minByOrNull { reading ->
                            val hour = Instant.ofEpochSecond(reading.timeEpochSeconds).atZone(zone).hour
                            kotlin.math.abs(hour - MIDDAY_HOUR)
                        } ?: readings.first()

                    ForecastDay(
                        date = date,
                        condition = WeatherCondition.entries
                            .getOrElse(representative.conditionId) { WeatherCondition.UNKNOWN },
                        description = representative.description,
                        iconCode = representative.iconCode,
                        minTemperatureCelsius = readings.minOf { it.temperatureCelsius },
                        maxTemperatureCelsius = readings.maxOf { it.temperatureCelsius },
                        precipitationProbability = readings.maxOf { it.precipitationProbability },
                        hourly = readings.map { reading ->
                            HourlyForecast(
                                time = Instant.ofEpochSecond(reading.timeEpochSeconds),
                                condition = WeatherCondition.entries
                                    .getOrElse(reading.conditionId) { WeatherCondition.UNKNOWN },
                                iconCode = reading.iconCode,
                                temperatureCelsius = reading.temperatureCelsius,
                                precipitationProbability = reading.precipitationProbability,
                                windSpeedMetresPerSecond = reading.windSpeedMetresPerSecond,
                            )
                        },
                    )
                }

        return Forecast(
            locationId = first.locationId,
            locationName = first.locationName,
            days = days,
            cachedAt = Instant.ofEpochSecond(first.cachedAtEpochSeconds),
            zoneOffset = offset,
        )
    }

    // ── SavedLocation ↔ Entity ─────────────────────────────────────────────

    fun SavedLocationEntity.toDomain(): SavedLocation = SavedLocation(
        id = id,
        name = name,
        countryCode = countryCode,
        state = state,
        latitude = latitude,
        longitude = longitude,
        sortOrder = sortOrder,
        isPrimary = isPrimary,
    )

    fun SavedLocation.toEntity(): SavedLocationEntity = SavedLocationEntity(
        id = id,
        name = name,
        countryCode = countryCode,
        state = state,
        latitude = latitude,
        longitude = longitude,
        sortOrder = sortOrder,
        isPrimary = isPrimary,
    )

    fun LocationSearchResult.toEntity(sortOrder: Int, isPrimary: Boolean): SavedLocationEntity = SavedLocationEntity(
        name = name,
        countryCode = countryCode,
        state = state,
        latitude = latitude,
        longitude = longitude,
        sortOrder = sortOrder,
        isPrimary = isPrimary,
    )

    /**
     * Builds a [ZoneOffset] from a raw seconds value, falling back to UTC.
     *
     * `ZoneOffset.ofTotalSeconds` throws outside ±18 hours. A malformed offset from the API, or
     * a corrupted cache row, must not crash the mapper, and UTC is the only defensible default.
     */
    private fun safeOffset(totalSeconds: Int): ZoneOffset =
        runCatching { ZoneOffset.ofTotalSeconds(totalSeconds) }.getOrDefault(ZoneOffset.UTC)

    /** Local noon: the hour whose reading best characterises a whole day. */
    private const val MIDDAY_HOUR = 12
}
